### ISSUE TITLE
Github Actions for sharing Kedro-Viz on Github Pages

### ISSUE THREAD

**Issue by NeroOkwa** [kedro-org/kedro-viz#1780]
## Description
This is related to #1574  and #1541 
For deploying your Kedro-Viz project to Github, implement a Github Action that bundles your build file, and transfers it to your Github repository.

## Context

This would further simplify (automate) the process of users deploying their Kedro-Viz visualisation to Github pages, and generating a shareable link. This link can then be shared with other team stakeholders for better collaboration.

## Possible Implementation
@deepyaman's example is a good starting point: [here](https://github.com/deepyaman/jaffle-shop/blob/main/.github/workflows/kedro-viz.yml)

## Acceptance Criteria
A user is able to easily host and share a version of their Kedro-Viz pipeline visualisation with a URL via Github Pages by:
1. Using the Github Action, build, transfer and deploy their kedro-viz project.
2. A unique Github URL is created with which they can access their project site.

**Comment by noklam** [comment]
https://github.com/ravi-kumar-pilla/shareableviz-action/blob/main/action.yml

Few comments @ravi-kumar-pilla:
- Should the Github action try to install python libraries requirements or kedro-viz. I think it should not try to do library installation. @astrojuanlu thoughts?
- telemetry may be a moving part , we may move away from .telemetry but it haven't happened yet.

**Comment by ravi-kumar-pilla** [comment]
Hi Team,

Posting some notes on publishing actions on GitHub here -

**Guide -** https://docs.github.com/en/actions/creating-actions/publishing-actions-in-github-marketplace

**Need for a new repository -** When you plan to publish your action to GitHub Marketplace, you'll need to ensure that the repository only includes the metadata file, code, and files necessary for the action. Creating a single repository for the action allows you to tag, release, and package the code in a single unit. GitHub also uses the action's metadata on your GitHub Marketplace page.

**Rules for the repo -**
* The action must be in a public repository.
* Each repository must contain a single action.
* Each repository must not contain any workflow files.
* The action's metadata file (action.yml or action.yaml) must be in the root directory of the repository.
* The name in the action's metadata file must be unique.
    * The name cannot match an existing action name published on GitHub Marketplace.
    * The name cannot match a user or organization on GitHub, unless the user or organization owner is publishing the action. For example, only the GitHub organization can publish an action named github.
    * The name cannot match an existing GitHub Marketplace category.
    * GitHub reserves the names of GitHub features.

**Example Action repos -**
1. https://github.com/atlanhq/dbt-action
2. https://github.com/marketplace/actions/trufflehog-oss
3. [More](https://github.com/marketplace?type=actions)

**Based on my understanding -**

1. We need to create a new repo under [KedroOrg](https://github.com/kedro-org) 

    **Possible repo names -** (Open for suggestions)
      1. kedro-viz-share-action
      2. kedro-viz-share
      3. shareableviz-action
      4. shareableviz
      5. kedro-viz-gh-pages
      6. viz-gh-pages-action

2. Create the repo similar to the example action repos with an action.yml file (testing this [here](https://github.com/ravi-kumar-pilla/shareableviz-action/blob/main/action.yml))

3. Branding an action -  We need to brand our action, but it seems to be a lengthy process and we should be a GitHub partner as per https://github.com/orgs/community/discussions/26767#discussioncomment-3253318. I am not sure if we already are ? cc: @NeroOkwa @astrojuanlu 

    I need help on this if we go the route of customizing the action logo. If not, we can choose from the available icons [here](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#branding). I need @stephkaiser inputs here. 

    Example : [cheatsheet](https://haya14busa.github.io/github-action-brandings/)
    ```
    branding:
      icon: share
      color: yellow
    ```

Thank you !

**Comment by ravi-kumar-pilla** [comment]
> https://github.com/ravi-kumar-pilla/shareableviz-action/blob/main/action.yml
> 
> Few comments @ravi-kumar-pilla:
> 
> * Should the Github action try to install python libraries requirements or kedro-viz. I think it should not try to do library installation. @astrojuanlu thoughts?

Without installing library requirements, as of now kedro-viz build will fail as it needs to parse datasets mentioned in the catalog. But I do agree with you that the action itself should only rely on kedro-viz and all the library specific installations should be part of the workflow rather than the action itself. We can document this and I would also like to hear from @astrojuanlu 

> * telemetry may be a moving part , we may move away from .telemetry but it haven't happened yet.

For the initial release of this action, I hope to keep it this way and if we move away from .telemetry, we need to think of updating this part

**Comment by astrojuanlu** [comment]
- I don't think we are technology partners yet, let's decouple this action from the branding since it's unclear what the timeline or requirements are to become one
- I am not an expert on GitHub Actions, could you look into 2 or 3 similar ones and see what they do to help us make the decision?

**Comment by stephkaiser** [comment]
re: branding/choosing the github action icon: 

unfortunately there is no diamond icon available in the feather icon library, and it also seems like it won't be possible to rotate a square icon by 45 degrees so I would agree the `share` icon is probably the most appropriate one to go with 👍 the yellow bg colour is also closest to our branding so lets go with that 

```
branding:
  icon: share
  color: yellow
```

**Comment by ravi-kumar-pilla** [comment]
Hi Team,

Thanks for joining the meeting and providing your suggestions. 

@astrojuanlu Regarding the option of choosing a custom GitHub action workflow as a publishing source for GitHub pages, we can handle that using an input. If users select this option, there will be no branch created and build artifacts will not be uploaded. More info [here](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site#publishing-with-a-custom-github-actions-workflow)

I implemented our action which can handle both the ways of configuring and enabling github pages. Please find the doc [here](https://github.com/ravi-kumar-pilla/shareableviz-action/blob/main/README.md). Please let me know if you have any questions.

Thank you

**Comment by ravi-kumar-pilla** [comment]
Hi Team,

I tested publishing the action to GitHub marketplace and the action works as expected - https://github.com/marketplace/actions/shareableviz-action

I have created a repo in KedroOrg as discussed - https://github.com/kedro-org/kedro-viz-share
and there is an initial PR for review which is a duplicate of the test repo published - https://github.com/kedro-org/kedro-viz-share/pull/1

I need some help from the team/admins -
1. The repo needs to be verified as per KedroOrg standards
2. Please comment on any required actions/security steps that need to be placed (example - branch protection, Code Quality workflows etc.)
3. I do not have access to run any github actions on this repo at this moment, so I was unable to run `test.yml`. It would be great to have this setup in.

<img width="1441" alt="image" src="https://github.com/kedro-org/kedro-viz/assets/87735534/dc509736-725b-414a-862c-6b84d465e9b2">
 
 cc: @rashidakanchwala @astrojuanlu @NeroOkwa 
 
Thank you

**Comment by astrojuanlu** [comment]
I don't have admin access to that repo either @merelcht

**Comment by merelcht** [comment]
FYI I enabled actions on this repo after chatting with @rashidakanchwala

**Comment by ravi-kumar-pilla** [comment]
> FYI I enabled actions on this repo after chatting with @rashidakanchwala

Hi @merelcht , thank you. I do have access now

**Comment by ravi-kumar-pilla** [comment]
Hi @merelcht ,

Can you please provide me the access to allow publishing the action to GitHub market place ?

![image](https://github.com/kedro-org/kedro-viz/assets/87735534/6686178e-c2f9-4332-9fd2-141b73c2448a)

Thank you

**Comment by merelcht** [comment]
@ravi-kumar-pilla Can you let me know what I should fill in here?
<img width="973" alt="Screenshot 2024-03-14 at 16 05 03" src="https://github.com/kedro-org/kedro-viz/assets/49397448/8cdb7f8e-98b3-4abd-9434-63e1d6f303a7">

**Comment by ravi-kumar-pilla** [comment]
Hi @merelcht , I see I have access now 👍 :) . But I do have a draft PR here - https://github.com/kedro-org/publish-kedro-viz/releases/edit/untagged-b93c404532ba486bfa80

**Comment by merelcht** [comment]
> Hi @merelcht , I see I have access now 👍 :) . But I do have a draft Release here - https://github.com/kedro-org/publish-kedro-viz/releases/edit/untagged-b93c404532ba486bfa80

Ahh okay I guess accepting the notice already unblocked you 👍
